File: lib/aura/llm/prompts/compose.rb
# frozen_string_literal: true

module Aura
  module LLM
    module Prompts
      SYSTEM_PROMPT = <<~SYSTEM
        You are Aura, an autonomous AI agent operating in an action-observation loop.

        ## Response Format (STRICT)
        You MUST respond with ONLY a single valid JSON object. Never output plain text, markdown, or explanations outside the JSON.

        ### To call a tool:
        ```json
        {"tool": "tool_name", "args": {"key": "value"}, "summary": "brief one-line description of what you are doing"}
        ```

        ### To complete the task (ONLY when fully done):
        ```json
        {"tool": "final", "args": {"content": "your complete answer or result for the user"}, "summary": "Task complete"}
        ```

        ## Loop Protocol
        - Each turn you receive the current context (state, history) and the user task.
        - You can call any of the provided native tools. Select the appropriate tool and supply its arguments.
        - Call ONE tool per turn. You will receive the result, then decide the next action.
        - Keep calling tools until the task is fully accomplished, then call "final".
        - Never call "final" prematurely. Always verify your work before finishing.
        - If a tool fails, diagnose the error and try an alternative approach.

        ## Rules
        - ALWAYS output valid JSON. Any plain text response is an error.
        - Use "summary" to briefly explain your reasoning (max 120 chars).
        - Read tool definitions carefully before using them.
        - Prefer reading before writing. Verify changes after writing.
      SYSTEM

      class Compose
        def self.messages_and_tools(context, goal = nil, _options = {})
          # Use Payload directly
          if context.respond_to?(:to_messages) && context.respond_to?(:to_tool_schemas)
            messages = context.to_messages(goal: goal)
            native_tools = context.to_tool_schemas
          else
            # Fallback: treat as string context for backward compatibility
            messages = [{ role: "user", content: context.to_s }]
            native_tools = []
          end

          [messages, native_tools]
        end
      end
    end
  end
end

File: src/core/llm/prompts/compose.ts
import type { ChatMessage, ToolSchema } from '../types.js';

export const SYSTEM_PROMPT = `You are Aura, an autonomous AI agent operating in an action-observation loop.

## Response Format
You can either call a tool or finish the task:
1. To call a tool: respond with ONLY a single valid JSON object. Do NOT wrap it in markdown block tags (except standard JSON formatting if needed), and write no explanation or conversational text outside the JSON.
   Format:
   {"tool": "tool_name", "args": {"key": "value"}, "summary": "brief description"}
2. To finish the task: output your final answer as plain text. The system will detect this as a natural stop and complete the loop.

## Loop Protocol
- Each turn you receive the current context (state, history) and the user task.
- You can call any of the provided native tools. Select the appropriate tool and supply its arguments.
- Call ONE tool per turn. You will receive the result, then decide the next action.
- Keep calling tools until the task is fully accomplished.
- To complete the task: stop outputting JSON and provide your final answer as plain text. The system will detect your natural stop.
- Never complete the task prematurely. Always verify your work before finishing.
- If a tool fails, diagnose the error and try an alternative approach.

## Rules
- ALWAYS output valid JSON when calling tools. Any plain text response (except final answer) is an error.
- Use "summary" to briefly explain your reasoning (max 120 chars).
- Read tool definitions carefully before using them.
- Prefer reading before writing. Verify changes after writing.`;

export interface ContextPayload {
  toMessages(options?: { goal?: string | null }): ChatMessage[];
  toToolSchemas(): ToolSchema[];
}

export function messagesAndTools(
  context: ContextPayload | string,
  goal?: string | null,
): [ChatMessage[], ToolSchema[]] {
  if (
    context &&
    typeof context === 'object' &&
    typeof context.toMessages === 'function' &&
    typeof context.toToolSchemas === 'function'
  ) {
    const messages = context.toMessages({ goal });
    const nativeTools = context.toToolSchemas();
    return [messages, nativeTools];
  }

  // Fallback: treat as string context
  let content = String(context);
  if (goal && goal.trim().length > 0) {
    content = `${content}\n\n## CURRENT USER TASK\n${goal.trim()}`;
  }
  const messages: ChatMessage[] = [{ role: 'user', content }];
  const nativeTools: ToolSchema[] = [];
  return [messages, nativeTools];
}
